File: Python/ejercicio_01.py
import itertools

import matplotlib.pyplot as plt

from graficas_util import carpeta_graficas, guardar_figura


def crear_graficas(espacio_muestral):
    out = carpeta_graficas(1)
    conteos = {0: 0, 1: 0, 2: 0, 3: 0}
    for resultado in espacio_muestral:
        conteos[resultado.count("C")] += 1
    etiquetas = [f"{k} caras" for k in conteos]
    plt.bar(etiquetas, list(conteos.values()), color=["#4c72b0", "#55a868", "#c44e52", "#8172b2"])
    plt.ylabel("Número de resultados")
    plt.title("Distribución del número de caras (3 monedas)")
    guardar_figura(out, "distribucion_caras")


def main():
    enunciado = """
=========================================================
EJERCICIO 1
Se lanzan tres monedas justas al aire.
a) Determine el espacio muestral del experimento.
b) ¿Cuál es la probabilidad de obtener exactamente dos caras?
c) ¿Cuál es la probabilidad de obtener al menos una cruz?
=========================================================
"""
    print(enunciado)

    # a) Espacio muestral
    print("--- SOLUCIÓN a) ---")
    print("Explicación: El espacio muestral de lanzar 3 monedas se obtiene combinando los resultados posibles (Cara 'C' o Cruz 'X') para cada una de las 3 posiciones.")
    moneda = ['C', 'X']
    espacio_muestral = list(itertools.product(moneda, repeat=3))
    print(f"-> Espacio muestral: {espacio_muestral}")
    print(f"-> Tamaño del espacio muestral: {len(espacio_muestral)}\n")

    # b) Probabilidad de obtener exactamente dos caras
    print("--- SOLUCIÓN b) ---")
    print("Explicación: Buscamos los eventos donde la 'C' aparece exactamente 2 veces. La probabilidad es la regla de Laplace (Casos Favorables / Casos Totales).")
    dos_caras = [resultado for resultado in espacio_muestral if resultado.count('C') == 2]
    prob_dos_caras = len(dos_caras) / len(espacio_muestral)
    print(f"-> Eventos favorables (exactamente 2 caras): {dos_caras}")
    print(f"-> Probabilidad = {len(dos_caras)} / {len(espacio_muestral)} = {prob_dos_caras:.4f}\n")

    # c) Probabilidad de obtener al menos una cruz
    print("--- SOLUCIÓN c) ---")
    print("Explicación: Buscamos eventos donde la 'X' aparece 1 o más veces. Esto es igual a decir que el resultado no sean todo caras (C,C,C).")
    al_menos_una_cruz = [resultado for resultado in espacio_muestral if resultado.count('X') >= 1]
    prob_al_menos_una_cruz = len(al_menos_una_cruz) / len(espacio_muestral)
    print(f"-> Probabilidad = {len(al_menos_una_cruz)} / {len(espacio_muestral)} = {prob_al_menos_una_cruz:.4f}")

    crear_graficas(espacio_muestral)


if __name__ == "__main__":
    main()

File: R/ejercicio_01.R
enunciado <- "
=========================================================
EJERCICIO 1 (R)
Se lanza un dado de 6 caras tres veces consecutivas.
a) ¿Cuántos resultados elementales tiene el espacio muestral?
b) Calcule la probabilidad de obtener una suma total de 10.
c) Calcule la probabilidad de que el primer lanzamiento sea mayor que el segundo y el segundo mayor que el tercero.
=========================================================
"
cat(enunciado)

cat("--- SOLUCIÓN a) ---\n")
cat("Explicación: Cada lanzamiento tiene 6 resultados posibles. Como son 3 lanzamientos independientes, usamos el principio multiplicativo: 6 * 6 * 6.\n")
resultados_totales <- 6^3
cat(sprintf("-> Resultados elementales: %d\n\n", resultados_totales))

cat("--- SOLUCIÓN b) ---\n")
cat("Explicación: Generamos el espacio muestral completo usando expand.grid y filtramos las combinaciones cuya suma sea 10.\n")
espacio <- expand.grid(d1=1:6, d2=1:6, d3=1:6)
suma_10 <- subset(espacio, d1 + d2 + d3 == 10)
casos_favorables_b <- nrow(suma_10)
prob_b <- casos_favorables_b / resultados_totales
cat(sprintf("-> Combinaciones que suman 10: %d\n", casos_favorables_b))
cat(sprintf("-> Probabilidad de suma 10 = %d / %d = %.4f\n\n", casos_favorables_b, resultados_totales, prob_b))

cat("--- SOLUCIÓN c) ---\n")
cat("Explicación: Buscamos combinaciones donde el primer dado sea mayor al segundo y este mayor al tercero (d1 > d2 > d3). Contamos los casos y dividimos por el total.\n")
mayor_estricto <- subset(espacio, d1 > d2 & d2 > d3)
casos_favorables_c <- nrow(mayor_estricto)
prob_c <- casos_favorables_c / resultados_totales
cat(sprintf("-> Combinaciones con d1 > d2 > d3: %d\n", casos_favorables_c))
cat(sprintf("-> Probabilidad = %d / %d = %.4f\n", casos_favorables_c, resultados_totales, prob_c))
